src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -1159,13 +1253,32 @@ public List<Pool> getSubPoolsForStackIds(Collection<Consumer> consumers, Collect
         return output;
     }
 
-    @SuppressWarnings("unchecked")
+    /**
+     * Retrieves pools that have a {@SourceSubscription} that has a source stack ID equal to the provided
+     * stack ID.
+     *
+     * @param owner
+     *  the owner of the pools to retrieve
+     *
+     * @param stackId
+     *  the ID of the stack for the source subscriptions to retrieve pools for
+     *
+     * @return the pools that have a source subscription with a source stack ID equal to the provided
+     *  stack ID.
+     */
     public List<Pool> getOwnerSubPoolsForStackId(Owner owner, String stackId) {
-        return createSecureCriteria()
-            .createAlias("sourceStack", "ss")
-            .add(Restrictions.eq("ss.sourceStackId", stackId))
-            .add(Restrictions.eq("owner", owner))
-            .list();
+        if (owner == null || stackId == null || stackId.isBlank()) {
+            return new ArrayList<>();
+        }
+
+        String query = "SELECT s.pool FROM SourceSubscription s " +
+            "WHERE s.pool.owner.id = :owner_id AND s.sourceStackId = :stack_id";
```

Comment:
So, this query wasn't even correct and there were no tests to expose that. For curiosity I corrected the JPQL to 
```
SELECT p 
FROM Pool p
WHERE p.owner.id = :owner_id AND p.sourceStack.sourceStackId = :stack_id
```

And this produced the following SQL:
```
select <Pool Fields> 
from cp_pool pool0_ 
cross join cp_pool_source_stack sourcestac1_ 
where pool0_.id=sourcestac1_.derivedPool_id and pool0_.owner_id=? and sourcestac1_.sourceStackId=?
```

However, I updated this function to use the JPA library to add the security predicate and I also added unit tests for this method, and the other changed methods!

Reply:
Cross join? Really, Hibernate? It doesn't matter since cross join + where clause against a join column like that _should_ result in the same query plan, but that is an interesting choice. Anyway, at least it's not doing a join to get to the owner ID and is looking at it directly on pool.